Pauli Twirling (PT) User Guide (#1848)
unitaryfund/mitiq: docs/source/guide/pt-4-low-level.md

---
jupytext:
  text_representation:
    extension: .myst
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.11.1
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

# What happens when I use PT?

The workflow of Pauli Twirling (PT) in Mitiq is represented in the figure below.

```{figure} ../img/pt_workflow.svg
---
width: 700px

> name: pt-workflow-overview

Reviewer: I think this might have to be renamed to something like `pt-workflow-overview2` because `pt-workflow-overview` is used in the main page `pt.md`.

Author: @purva-thakre That's by design, since we're using the same image for both places in the docs. Take a look [here](https://mitiq--1848.org.readthedocs.build/en/1848/guide/pt-4-low-level.html) and we can see the code execution errors.

That said, I keep finding inaccuracies (both in the code and docs/theory), so this one isn't quite ready for review yet (when I marked it as so late last night I was apparently rather hopeful 🤣)!

Author: Thanks @purva-thakre, should be all set now!

---
jupytext:
  text_representation:
    extension: .myst
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.11.1
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

# What happens when I use PT?

The workflow of Pauli Twirling (PT) in Mitiq is represented in the figure below.

```{figure} ../img/pt_workflow.svg
---
width: 700px
name: pt-workflow-overview-2
---
Workflow of the PT technique in Mitiq, detailed in the [What happens when I use PT?](pt-4-low-level.md) section.
```

- The user provides a `QPROGRAM`, (i.e. a quantum circuit defined via any of the supported [frontends](frontends-backends.md)).
- Mitiq modifies the input circuit with the insertion of PT gates on noisy operations.
- The modified circuit is executed via a user-defined [Executor](executors.md).
- The error mitigated expectation value is returned to the user.

With respect to the workflows of other error-mitigation techniques (e.g. [ZNE](zne-4-low-level.md) or [PEC](pec-4-low-level.md)),
PT involves the generation of a _single_ randomly-modified circuit, and then averages over its executions.

> For this reason, there is no need for a complex final inference step, which is necessary for other 
> techniques, and so this average is instead trivial for PT.

Reviewer 2: ```suggestion
For this reason, there is no need for a complex final inference step (unlike in other 
techniques), and so this average is instead trivial for PT.
```

Pauli Twirling doesn't require running the circuit at different noise levels or with different noise models. It applies a randomized sequence of Pauli operations within the same quantum circuit and averages the results to reduce the effect of the noise.

```{code-cell} ipython3
import numpy as np
from typing import cast

average = cast(float, np.average(expvals))
print(average)
```

Keep in mind, ths code is for illustration and that the noise level, type of noise (here amplitude damping), and the observable need to be adapted to the specific experiment.

If executed on a noiseless backend, a given `circuit_with_pt` and `circuit` are equivalent.
On a real backend, they have a different sensitivity to noise. The core idea of the PT technique is that,
`circuits_with_pt` (hopefully) tailors the noise into stochastic Pauli channels, such that a simple average over results
will return a mitigated result.

As a final remark, we stress that the low-level procedure that we have shown is exactly what {func}`.execute_with_pt()` does behind the scenes.
Let's verify this fact: 

```{code-cell} ipython3
np.isclose(
  pt.execute_with_pt(circuit, executor),
  average,
)

> ```

Reviewer 2: Unfortunately, the results displayed don't match the text- `np.isclose` asserts `False`, see https://mitiq--1848.org.readthedocs.build/en/1848/guide/pt-4-low-level.html .

```{code-cell} ipython3
from mitiq import pt

circuit_to_twirl = circuit.copy()
CNOT_twirled_circuits = pt.twirl_CNOT_gates(circuit_to_twirl, num_circuits=10)
twirled_circuits = [
    pt.twirl_CZ_gates(c, num_circuits=1)[0] for c in CNOT_twirled_circuits
]
print(CNOT_twirled_circuits[0:1])
print(twirled_circuits[0:1])
```
~~We see that we return lists of the randomly twirled circuits, and so we must take a simple average over their expectation values.~~

## Verifying Pauli Twirling Works

It is worth checking if the technique does tailor the noise as expected. For this
demonstration, we compare the off-diagonal terms in the Pauli Transfer Matrix (PTM)
{cite}`Hashim_2021_PRX` of the original input circuit, the noisy input circuit as well
as the averaged PTM of the different possible Pauli twirled circuits. A detailed discussion
on the PTM is available in [theory section of PT](pt-5-theory.md).

Now, instead of using the ideal input circuit, we use PT on the noisy circuit with the
default option for the number of twirled circuits. 

```{code-cell} ipython3

> from mitiq import pt

Reviewer: ```suggestion
from mitiq import pt
from cirq import amplitude_damp
from mitiq.interface import convert_to_mitiq

noise_level = 0.1
mitiq_circuit, _ = convert_to_mitiq(circuit)
noisy_circuit = mitiq_circuit.with_noise(amplitude_damp(gamma=noise_level))
```

tailored_circuit_noisy = pt.pauli_twirl_circuit(
    circuit = noisy_circuit
)
```
The PTM of the ideal circuit is in the following code block. 

The PTM of the noisy circuit is in the following code block. 

The PTM of the output from PTM is in the following code block. 

~~Pauli Twirling doesn't require running the circuit at different noise levels or with different noise models. It applies a randomized sequence of Pauli operations within the same quantum circuit and averages the results to reduce the effect of the noise.~~

```{code-cell} ipython3
import numpy as np
from typing import cast

average = cast(float, np.average(expvals))
print(average)
```

Keep in mind, ths code is for illustration and that the noise level, type of noise (here amplitude damping), and the observable need to be adapted to the specific experiment.

> On a real backend, they have a different sensitivity to noise. The core idea of the PT technique is that,

Reviewer: If one were to demonstrate an example of PT + some other mitigation technique, is it preferable to 

1. introduce a separate tutorial?
2. Or demonstrate it in this section to keep it consistent with what @Aaron-Robertson originally intended?
